File: frontend/de/unika/ipd/grgen/ast/stmt/DefDeclStatementNode.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 7.1
 * Copyright (C) 2003-2025 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */
package de.unika.ipd.grgen.ast.stmt;

import java.util.Collection;
import java.util.Vector;

import de.unika.ipd.grgen.ast.*;
import de.unika.ipd.grgen.ast.decl.DeclNode;
import de.unika.ipd.grgen.ast.decl.pattern.ConstraintDeclNode;
import de.unika.ipd.grgen.ast.decl.pattern.EdgeDeclNode;
import de.unika.ipd.grgen.ast.decl.pattern.NodeDeclNode;
import de.unika.ipd.grgen.ast.decl.pattern.VarDeclNode;
import de.unika.ipd.grgen.ast.pattern.ConnectionNode;
import de.unika.ipd.grgen.ast.pattern.SingleNodeConnNode;
import de.unika.ipd.grgen.ast.util.Resolver;
import de.unika.ipd.grgen.ir.stmt.DefDeclGraphEntityStatement;
import de.unika.ipd.grgen.ir.stmt.DefDeclVarStatement;
import de.unika.ipd.grgen.ir.IR;
import de.unika.ipd.grgen.ir.pattern.GraphEntity;
import de.unika.ipd.grgen.ir.pattern.Variable;
import de.unika.ipd.grgen.parser.Coords;

/**
 * AST node representing a def declaration statement node (a variable that can be assigned in an attribute evaluation statements).
 */
public class DefDeclStatementNode extends EvalStatementNode
{
	static {
		setName(DefDeclStatementNode.class, "def decl statement");
	}

	BaseNode defDeclUnresolved;
	int context;

	VarDeclNode defDeclVar;
	ConstraintDeclNode defDeclGraphElement;

	public DefDeclStatementNode(Coords coords, BaseNode target, int context)
	{
		super(coords);
		this.defDeclUnresolved = target;
		becomeParent(this.defDeclUnresolved);
		this.context = context;
	}

	public DefDeclStatementNode(Coords coords, VarDeclNode defDeclVar, int context)
	{
		super(coords);
		this.defDeclVar = defDeclVar;
		this.context = context;
	}

	/** returns children of this node */
	@Override
	public Collection<BaseNode> getChildren()
	{
		Vector<BaseNode> children = new Vector<BaseNode>();
		children.add(getValidVersion(defDeclUnresolved, defDeclVar, defDeclGraphElement));
		return children;
	}

	/** returns names of the children, same order as in getChildren */
	@Override
	public Collection<String> getChildrenNames()
	{
		Vector<String> childrenNames = new Vector<String>();
		childrenNames.add("defDecl");
		return childrenNames;
	}

	/** @see de.unika.ipd.grgen.ast.BaseNode#resolveLocal() */
	@Override
	protected boolean resolveLocal()
	{
		boolean successfullyResolved = true;
		DeclNode decl = getDecl();
		if(decl.typeUnresolved instanceof PackageIdentNode)
			Resolver.resolveOwner((PackageIdentNode)decl.typeUnresolved);
		else
			fixupDefinition(decl.typeUnresolved, decl.typeUnresolved.getScope());
		successfullyResolved = decl.resolve();
		return successfullyResolved;
	}

	@Override
	protected boolean checkLocal()
	{
		return true;
	}

	@Override
	public boolean checkStatementLocal(boolean isLHS, DeclNode root, EvalStatementNode enclosingLoop)
	{
		return true;
	}

	public DeclNode getDecl()
	{
		if(defDeclUnresolved == null)
			return defDeclVar;
		
		if(defDeclUnresolved instanceof VarDeclNode) {
			defDeclVar = (VarDeclNode)defDeclUnresolved;
			return defDeclVar;
		} else if(defDeclUnresolved instanceof SingleNodeConnNode) {
			SingleNodeConnNode sncn = (SingleNodeConnNode)defDeclUnresolved;
			defDeclGraphElement = (NodeDeclNode)sncn.nodeUnresolved;
			return defDeclGraphElement;
		} else if(defDeclUnresolved instanceof ConstraintDeclNode) {
			defDeclGraphElement = (ConstraintDeclNode)defDeclUnresolved;
			return defDeclGraphElement;
		} else {
			ConnectionNode cn = (ConnectionNode)defDeclUnresolved;
			defDeclGraphElement = ((EdgeDeclNode)cn.edgeUnresolved);
			return defDeclGraphElement;
		}
	}

	@Override
	protected IR constructIR()
	{
		// potential initialization is attached to the Var or the GraphEntity
		if(defDeclVar != null) {
			Variable var = defDeclVar.checkIR(Variable.class);
			return new DefDeclVarStatement(var);
		} else {
			GraphEntity graphEntity = defDeclGraphElement.checkIR(GraphEntity.class);
			return new DefDeclGraphEntityStatement(graphEntity);
		}
	}
}

File: frontend/de/unika/ipd/grgen/ast/stmt/DefDeclStatementNode.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>
namespace de.unika.ipd.grgen.ast.stmt
{

using System.Collections.Generic;

using de.unika.ipd.grgen.ast;
using DeclNode = de.unika.ipd.grgen.ast.decl.DeclNode;
using ConstraintDeclNode = de.unika.ipd.grgen.ast.decl.pattern.ConstraintDeclNode;
using EdgeDeclNode = de.unika.ipd.grgen.ast.decl.pattern.EdgeDeclNode;
using NodeDeclNode = de.unika.ipd.grgen.ast.decl.pattern.NodeDeclNode;
using VarDeclNode = de.unika.ipd.grgen.ast.decl.pattern.VarDeclNode;
using ConnectionNode = de.unika.ipd.grgen.ast.pattern.ConnectionNode;
using SingleNodeConnNode = de.unika.ipd.grgen.ast.pattern.SingleNodeConnNode;
using de.unika.ipd.grgen.ast.util;
using DefDeclGraphEntityStatement = de.unika.ipd.grgen.ir.stmt.DefDeclGraphEntityStatement;
using DefDeclVarStatement = de.unika.ipd.grgen.ir.stmt.DefDeclVarStatement;
using IR = de.unika.ipd.grgen.ir.IR;
using GraphEntity = de.unika.ipd.grgen.ir.pattern.GraphEntity;
using Variable = de.unika.ipd.grgen.ir.pattern.Variable;
using Coords = de.unika.ipd.grgen.parser.Coords;

/// <summary>
/// AST node representing a def declaration statement node (a variable that can be assigned in an attribute evaluation statements).
/// </summary>
public class DefDeclStatementNode : EvalStatementNode
{
	static DefDeclStatementNode()
	{
		SetClassName(typeof(DefDeclStatementNode), "def decl statement");
	}

	internal BaseNode defDeclUnresolved;
	internal int context;

	internal VarDeclNode defDeclVar;
	internal ConstraintDeclNode defDeclGraphElement;

	public DefDeclStatementNode(Coords coords, BaseNode target, int context)
		: base(coords)
	{
		this.defDeclUnresolved = target;
		BecomeParent(this.defDeclUnresolved);
		this.context = context;
	}

	public DefDeclStatementNode(Coords coords, VarDeclNode defDeclVar, int context)
		: base(coords)
	{
		this.defDeclVar = defDeclVar;
		this.context = context;
	}

	/// <summary>
	/// returns children of this node </summary>
	public override ICollection<BaseNode> Children
	{
		get
		{
		IList<BaseNode> children = new List<BaseNode>();
		children.Add(GetValidVersion(defDeclUnresolved, defDeclVar, defDeclGraphElement));
		return children;
		}
	}

	/// <summary>
	/// returns names of the children, same order as in getChildren </summary>
	public override ICollection<string> ChildrenNames
	{
		get
		{
		IList<string> childrenNames = new List<string>();
		childrenNames.Add("defDecl");
		return childrenNames;
		}
	}

	/// <seealso cref="de.unika.ipd.grgen.ast.BaseNode.resolveLocal() "/>
	protected internal override bool ResolveLocal()
	{
		bool successfullyResolved = true;
		DeclNode decl = Decl;
		if(decl.typeUnresolved is PackageIdentNode)
			Resolver.ResolveOwner((PackageIdentNode)decl.typeUnresolved);
		else
			FixupDefinition(decl.typeUnresolved, decl.typeUnresolved.Scope);
		successfullyResolved = decl.Resolve();
		return successfullyResolved;
	}

	protected internal override bool CheckLocal()
	{
		return true;
	}

	public override bool CheckStatementLocal(bool isLHS, DeclNode root, EvalStatementNode enclosingLoop)
	{
		return true;
	}

	public virtual DeclNode Decl
	{
		get
		{
		if(defDeclUnresolved == null)
			return defDeclVar;

		if(defDeclUnresolved is VarDeclNode)
		{
			defDeclVar = (VarDeclNode)defDeclUnresolved;
			return defDeclVar;
		}
		else if(defDeclUnresolved is SingleNodeConnNode)
		{
			SingleNodeConnNode sncn = (SingleNodeConnNode)defDeclUnresolved;
			defDeclGraphElement = (NodeDeclNode)sncn.nodeUnresolved;
			return defDeclGraphElement;
		}
		else if(defDeclUnresolved is ConstraintDeclNode)
		{
			defDeclGraphElement = (ConstraintDeclNode)defDeclUnresolved;
			return defDeclGraphElement;
		}
		else
		{
			ConnectionNode cn = (ConnectionNode)defDeclUnresolved;
			defDeclGraphElement = ((EdgeDeclNode)cn.edgeUnresolved);
			return defDeclGraphElement;
		}
		}
	}

	protected internal override IR ConstructIR()
	{
		// potential initialization is attached to the Var or the GraphEntity
		if(defDeclVar != null)
		{
			Variable var = defDeclVar.CheckIR(typeof(Variable));
			return new DefDeclVarStatement(var);
		}
		else
		{
			GraphEntity graphEntity = defDeclGraphElement.CheckIR(typeof(GraphEntity));
			return new DefDeclGraphEntityStatement(graphEntity);
		}
	}
}

}
